### ISSUE TITLE
Advanced production build

### ISSUE THREAD

**Issue by mgechev** [mgechev/angular2-seed#864]
Currently our production build is quite basic and produces only two bundles:

- shims
- libs + app

Which means that we're not taking advantage of:

- lazy loading
- offline template compilation
- tree-shaking

## Proposal

Incrementally improve the production build by:

### Step 1

[ ] - Include offline template precompilation
[ ] - Introduce tree-shaking with rollup.js

### Step 2

[ ] - Introduce bundling by bounded context (lazy-loading boundary) strategy

**Comment by hermagrini** [comment]
Would Critical Path CSS be something interesting to generate as well?

**Comment by phiphou** [comment]
As I said in #860 , application (only the code you/we wrote), vendors (angular2, rxjs and potential additional vendors like jquery, lodash...), and polyfills (reflect-metadata, zone, es6-shim,...) life-cycles aren't the same most of time. Keeping them in separate bundles seems to me a good practice. I'll be interested in knowing your thoughts about this point 

I also agree lazy loading would be a great feature. It would  obviously augment the amount of generated bundles, but I don't see any major reason not to working this way.

**Comment by Bigous** [comment]
Lazy loading and hot loading are amazing. We could create a pattern ... I though, when it was changed, that components that started with + in it's directory would be lazy loaded only when needed. I like this pattern, what do you think?

**Comment by jlberrocal** [comment]
@Bigous i'm using the + for indicate that is a lazy loaded component, for development seems to work fine, but for production there is still only 2 JS files, one with the shims and a big one with the vendors and the app itself

**Comment by Bigous** [comment]
Yes thats why I think we could create a name pattern for that. Bundle the components in other files, like AngularClass did with the webpack starter kit would be a nice one and would resolve the step 2. Unfortunately I've got almost no time during the past 3 months to help with the project, but the idea is still very valid.

**Comment by mgechev** [comment]
[Here's a blog post I wrote](http://blog.mgechev.com/2016/06/26/tree-shaking-angular2-production-build-rollup-javascript/) about an experiment I did around the production build efficiency.

A working "Hello world!" example with precompilation + tree-shaking, can be found [here](https://github.com/mgechev/angular2-ngc-rollup-build).

**Comment by mgechev** [comment]
In addition - I'll be playing with the production build to introduce:

- Brotli
- Google Clojure Compiler

These tools should decrease the bundle size to around 20k.

**Comment by Bigous** [comment]
20k footprint for ng2 is unbelievable! Awesome... and congratulations for the great article!

**Comment by vyakymenko** [comment]
@mgechev , I'm not sure about `Google Clojure Compiler`, because there are so many real testing information about `GCC vs UglyfyJS2`.

https://www.peterbe.com/plog/advanced-closure-compiler-vs-uglifyjs2

**Comment by mgechev** [comment]
Update: It is still too early to use offline compilation. The platform directives need to be explicitly configured and also I had issues making it work when `ngModel` directive is used.

In general, my plan is to implement a task called `build.prod.exp` which is going to be an experimental production build which includes everything from above.

**Comment by SekibOmazic** [comment]
@mgechev When could we expect tree-shaking with rollup.js? Any timeline on this?

**Comment by mgechev** [comment]
@SekibOmazic there's WIP in this branch https://github.com/mgechev/angular2-seed/tree/advanced-build

**Comment by SekibOmazic** [comment]
@mgechev cool! Hey did you just add rollup.js?

**Comment by mgechev** [comment]
Yes only rollup. There are some rxjs resolution issues that need to be addressed + refactoring, before being able to merge in master.

**Comment by SekibOmazic** [comment]
@mgechev running `gulp build.prod.exp` creates a `tmp/app.js` of 1.6M. It doesn't look really "squeezed". Am I missing something?

**Comment by mgechev** [comment]
There's no minification and compression of the code, as I already said the work in this branch is Work In Progress.

**Comment by ZuSe** [comment]
I also would like to see support for cloud foundry/heroku deployment.
Right now a lot of unnecessary files is shipped.

**Comment by Perezmarc** [comment]
Are you supporting gzipped build at the moment? my app.js file is 2.5mb and my project is not that huge. On production it is 1.6 s to load for the first time...

**Comment by vyakymenko** [comment]
@Perezmarc , gzipped build o_O, what do you mean when you talking about `gzipped build` ?

Gzip is using on `server-side`.

**Comment by Perezmarc** [comment]
sorry for my ignorance, looked at the post http://blog.mgechev.com/2016/06/26/tree-shaking-angular2-production-build-rollup-javascript/ and as build.prod currently generates a minified code, I thought that the next step for reducing the load time was gzipping it. :S I can't generate then a build to deploy it to S3 gzipped? In order to gzip it, what should i do?

**Comment by vyakymenko** [comment]
@Perezmarc , for example, if you using NginX for representing your application, you can look for my example for configuring `gzip compression` [here](https://github.com/vyakymenko/angular2-seed-express#how-to-configure-my-nginx).

But, I'm still confusing, what do you mean about deploy `gzipped code`. I think you need more documentation and literature about how it's work.

**Comment by Perezmarc** [comment]
I thought a gzipped code could be deployed somehow and reduce load time, but probably it makes no sense... Probably my only option is to use angular-universal and render on server to reduce that 1.6 seconds? and lazy loading.

**Comment by mgechev** [comment]
FYI I just added experimental support of AoT compilation support https://github.com/mgechev/angular2-seed/commit/bf5a3322329ecc603ea97d101da79f789cb3d70b. It'll be great to test it and once we feel that it is stable enough, merge with the master branch.

**Comment by mgechev** [comment]
Also keep in mind that it may not work on windows https://github.com/angular/angular/issues/8582.

**Comment by SekibOmazic** [comment]
@mgechev It works fine on Mac, no problems so far. Bundle size is now 1.1M (minified but not gzipped) which is still quite big. Will this be further optimised?

Another proposal: create 3 bundles in prod build: shims, vendor (angular) and app.

**Comment by mgechev** [comment]
The ahead of time compilation doesn't reduce the bundle size. Once the seed includes Google Closure Compiler we can drop the size more than 10 times, thanks fo tree-shaking and compression (compression we can use today as well).

**Comment by sfabriece** [comment]
Shouldn't it reduce size due to the compiler not being included?

**Comment by vyakymenko** [comment]
@sfabriece , don't forget about gzip.

**Comment by mgechev** [comment]
@sfabriece correct, which is still significant improvement. By this we get 0.5M off - 1.6M -> 1.1M.

Today I'll work on integration of Google Closure Compiler which should reduce the size even more, although we won't be able to perform tree-shaking yet.

**Comment by mgechev** [comment]
FYI we now have Google Closure Compiler support https://github.com/mgechev/angular2-seed/commit/e8105f1627a752b2bae2d1d575a65beecf982c8c. I'd recommend you to give it a try and let me know if you face any issues. The entire JavaScript payload now is ~170K and will drop dramatically once GCC supports `export *`.

**Comment by mpetkov** [comment]
Hey all, I just read through this entire thread and other threads about lazy-loading/async routing.  Wanted to find out if there is any timeframe on when it will be implemented? 

I am currently working on application that has a single codebase but many seprate screens (routes).  Currently the one prod bundle is getting too big so I am forced to create separate projects for each route to improve loading performance.

**Comment by mgechev** [comment]
@mpetkov the first step is to reduce the size of the production build. Currently it's 150k (gzipped), it should drop to ~50-70K once we get the tree-shaking work properly.

**Comment by ruffiem** [comment]
@mgechev the output size is quite big indeed... What is wrong with the tree-shaking ?

**Comment by mgechev** [comment]
We can provide solution for tree-shaking which works in the general case because a lot of projects use commonjs modules. Although there's a commonjs plugin for rollup it might get tricky to configure it for the variety of packages available out there. Google Closure Compiler seems like a good option which can provide bundling, minification and tree-shaking in a single step but we're still blocked by their lack of support for `export *`.

**Comment by ruffiem** [comment]
So either we should define a standard use of commonjs modules or we wait for the Closure Compiler to implement the export wildcard. Just ran a few test using the Closure Compiler and faced `input0:2: ERROR - ES6 transpilation of 'Wildcard export' is not yet implemented.`. But that's good news, it says 'not yet implemented'.

**Comment by ruffiem** [comment]
We'll maybe get something before the end of the year... https://github.com/google/closure-compiler/issues/878

**Comment by mo-ba** [comment]
imho tree shaking is a nice (to have) feature, but in terms of speed lazy loading gives a bigger boost, especially as applications get bigger...
is there any reason why you want to implement it in that order?

**Comment by mgechev** [comment]
@mo-ba you can take a look at this issue. Lazy-loading seems possible at this point https://github.com/mgechev/angular2-seed/issues/1358.

**Comment by mgechev** [comment]
Today I did some experiments with rollup.js and was able to reduce the size of the production bundle from 131K to 86K (gzipped & minified). At first this sounds really attractive, however, the consumption of cjs modules can get tricky.

My suggestion is to implement an experimental bundling with tree-shaking, taking advantage of rollup and keeping the default `build.prod` & `build.prod.exp` the way they are. This way only people who are aware what they do will be able to reduce the bundle size of their apps even further.

What do you think?

**Comment by Bigous** [comment]
Hi @mgechev, Ionic v2 RC0 is using rollup and it's a pain to add external libraries. Clojure was our another option... I like to keep the build.prod in safe mode :D and add a build.prod.exp with rollup or clojure.

**Comment by ruffiem** [comment]
@mgechev - If Closure cause less trouble consuming cjs modules, we should maybe wait until the export wildcard is available. In the meantime, experimenting around rollup is very appreciated though : 86k is sick but what does it include ?

@Bigous - what kind of support do we have to import external libs with rollup ? rollup keeps them out the bundle ?

**Comment by Bigous** [comment]
@ruffiem If the library was not developed to be tree shake-able, like libraries that are just a wrapper to other library - lodash, highcharts, charts, etc... - you need to provide a custom rollup settings. But as it is now, it's just an npm install and you are good to go.
We already have to do that in our seed, because we use SystemJS, so... it would not be a game break, but I like to keep things simple. You can see the ionic discussion driftyco/ionic-app-scripts#68 . It was closed, but still not resolved and people are moving back to version beta.11 so they can avoid rollup config...

**Comment by mgechev** [comment]
FYI I'm experimenting with lazy loading here https://github.com/mgechev/angular-seed/tree/lazy so we may have something working by the end of the week.

**Comment by RoxKilly** [comment]
About Closure's incompatibility: Angular gave a presentation at AngularConnect that suggested [`tsickle`](https://github.com/angular/tsickle) as a bridge to transpile `ng2 typescript` to Closure compatible JS. [See this video](https://www.youtube.com/watch?v=_VnV7R4Ncwc) from 23:15 to 24:10.

Is that info relevant to this discussion?

**Comment by mgechev** [comment]
Yes, I think running closure build with preprocessed with tsickle code is possible. I have played with tsickle recently and given that it's wrapper around tsc the integration with the seed should be relatively straightforward.

Here's a post about my experience http://blog.mgechev.com/2016/07/21/even-smaller-angular2-applications-closure-tree-shaking/.

In the "lazy" branch I am also playing with lazy loading which given the bundle arithmetic of SystemJS builder should not be that hard to implement neither. I am not sure how much time I will be able to spend on this in the next a couple of days/by the end of the month.

If anyone is interested in implementing any of these tasks we can chat and I can share some pointers.

**Comment by mgechev** [comment]
Btw, from both tasks, lazy loading seems higher priority, easier to implement and won't vary much in near future.

**Comment by SamVerschueren** [comment]
I'm looking into the lazy loading part with what you have done so far. I always get this error message

> Cannot find 'default' in 'app/+foo/foo.module.js'

I also tried it without `export default` and attaching `#FooModule` at the end of the path, but still no dice. Anyone who knows more about this issue?

**Comment by RoxKilly** [comment]
@SamVerschueren do you have an import statement without brackets anywhere? eg:
```
import MyModule from 'app/my.module';
```
Should be
```
import {MyModule} from 'app/my.module';
```

**Comment by SamVerschueren** [comment]
@RoxKilly With lazy loading, you don't have things like `import * from 'app/my.module'` because that will actually already import the code and that would mean you can't break those pieces into 2.

This means you have something like this

```ts
@NgModule({})
export default FooModule { }
```

```ts
const routes: Routes = [
	{ path: 'foo', loadChildren: 'app/+foo/foo.module' }
];
```

This results in the error

> Cannot find 'default' in 'app/+foo/foo.module.js'

Someone at the Angular2 gitter said that `export default` does not work with AoT. Not sure about that though, but in order to make sure it works I tried it like this.

```ts
@NgModule({})
export FooModule { }
```

```ts
const routes: Routes = [
	{ path: 'foo', loadChildren: 'app/+foo/foo.module#FooModule' }
];
```

Which results in 

> Unhandled Promise rejection: No NgModule metadata found for 'FooModule'. ; Zone: angular ; Task: Promise.then

I already read through [this thread](https://github.com/angular/angular/issues/10933) and tried most of the suggestions, no dice.

**Comment by RoxKilly** [comment]
@SamVerschueren 

> export FooModule { }

Should be
```
export class FooModule {}
```

**Comment by SamVerschueren** [comment]
Jup sorry, was just a typo here. I have all that and the lazy loading works in development so that doesn't seem to be the problem. The problem is when building the bundles.

Apparently SystemJS has a `bundles` option in the config. But, by default, angular-seed does not provide a config for production build as it is a static build. So I started adding the config directly in `index.html` with the `bundles` property and I get this far that it actually succesfully loads the module and shows this error

> Uncaught (in promise): Error: No NgModule metadata found for 'FooModule'.

I might be getting closer to a solution but do I really have to generate that `metadata.json` file per bundle? Not sure how to do that either so if someone else could point me in the right direction...

**Comment by RoxKilly** [comment]
@SamVerschueren  Oh I see. it's only when you bundle that it stops working.  Then you've made it further than I. I gave up on lazy loading (for now) because I couldn't get it to work with my bundling build.

**Comment by doapp-ryanp** [comment]
Is it possible to gzip all `.html,.css,.js` as part of `build.prod` so they can be directly uploaded to static object store (s3+CloudFront)? 

We don't support web clients that don't support `Accept-Encoding: gzip`

**Comment by SamVerschueren** [comment]
@doapp-ryanp Seems like an "easy" extra step that can be written as custom task, no?

**Comment by doapp-ryanp** [comment]
Yup just didn't know if was already build in (and I was just missing a flag/opt). Seems like a very common usage pattern.  I will make a PR at some point to add.

**Comment by mgechev** [comment]
Hey @doapp-ryanp, I think @SamVerschueren has a point. Lets keep this out of the seed, and maybe write a blog post/wiki article. I recently got rid of 20% of the dependencies and I'm willing to keep this process of shrinking the seed so lets not introduce new dependencies for now.

**Comment by SamVerschueren** [comment]
@mgechev If you have any docs that I could use in order to make my SystemJS build work with lazy loading, feel free to provide that information. I'm really struggling for hours, if not days, with this problem. Would love to have this in the seed by default.

**Comment by mgechev** [comment]
@SamVerschueren [here's a demo](https://github.com/angular/angular/blob/master/modules/playground/src/routing/app/inbox-app.ts) in the official Angular repo. If it still doesn't work after that we can dig deeper into it.

**Comment by SamVerschueren** [comment]
The thing is that it works while developing (`npm start`). But the moment I try to create self executing bundles, something fails and I just can't get it to work.

The question is, can we still use `builder.buildStatic` here or should we use `builder.bundle` together with a `System.import` in `index.html` instead.

**Comment by mgechev** [comment]
From what I left in the lazy branch I remember that I had to exclude Angular from the production lazy loaded bundles. I think this might have led to some issues in non-recognizing metadata. I've had this issue in the past when I run Angular Universal in Service Workers.

**Comment by scaryroy** [comment]
@mgechev so are you saying your lazy branch will work if we dont use export *?

**Comment by mgechev** [comment]
No, the lazy branch doesn't use Google Closure Compiler.

**Comment by mgechev** [comment]
@SamVerschueren here's a [working version](https://github.com/mgechev/angular-seed/tree/lazy). Next step is to make it work with AoT.

**Comment by SamVerschueren** [comment]
@mgechev You are a hero! Thanks for looking into this and finding what caused the issues. I didn't tried it yet but I will when I find some more time.

**Comment by ruffiem** [comment]
@mgechev I can't make external libraries work in prod on the lazy branch, is that normal ?

**Comment by mgechev** [comment]
The [lazy branch](https://github.com/mgechev/angular-seed/tree/lazy) now supports lazy loading with both AoT and JiT compilation. The code there is just proof-of-concept at this point but most likely I'll work on some advanced build in the next a couple of days.

**Comment by SamVerschueren** [comment]
I thought creating a new issue for this but it makes more sense to put things here.

I wanted to add treeshaking to my project which started from this seed. So I gave it a go and had something working fairly quickly. BUT, I was importing some CommonJS projects from npm in the following way.

```js
import * as x from 'x';

x();
```

This is not allowed for Rollup as it will fail with something like `Could not call namespace x`. So I started investigating this. I got it working by changing that import statement to `import x from 'x';` and set the `allowSyntheticDefaultImports` to `true` in the `tsconfig.json`. Production build with AoT and rollup was working perfectly!

But when I ran `npm start` again, it failed with something like `could not find x_1.default` as I was now importing the default method. I started to investigate this and came across the inter-format dependencies in the SystemJS docs https://github.com/systemjs/systemjs/blob/master/docs/module-formats.md#inter-format-dependencies. So why wasn't that working for me? I thought SystemJS would hijack the `require` function and exposes my CJS module as `default`.

I opened an issue https://github.com/systemjs/systemjs/issues/1502 and got a response. Apparently, for the inter format dependencies to work, you have to transpile your code to `system` instead of `commonjs`. But with this came a 3rd issue, `module.id` does not work anymore as it is CommonJS specific (see issue). I had to change all my `module.id` calls to `__moduleName`.

Those steps where required to get Rollup working together with a working dev build.

- Change all the `import * as` statements to `import` statements
- Transpile to `system` instead of `commonjs`
- Change `module.id` to `__moduleName`

Quite some changes though, but this got everything working.

@mgechev Yes SystemJS uses rollup, but not sure how good that works. My build with SystemJS was 1.8 MB and when I used rollup directly, I had 900 kB.

**Comment by mgechev** [comment]
Thanks for the comment! It's great to have this information logged somewhere.

I've considered switching to plain Rollup from SystemJS, however, I'm not sure it's a good idea to move drop SystemJS yet. A lot of people are using plenty of third party CommonJS dependencies and the configuration of rollup can become very complicated.

I've been thinking of having a mixed approach where we support bundling with both SystemJS and rollup. Rollup can be a separate task and we can switch to this build by having something like:

```shell
npm run build.prod.exp -- --rollup
npm run build.prod -- --rollup
```

On the other hand, having tsickle with Google Closure Compiler will produce even smaller bundle and will do everything (bundling, tree-shaking, minification) in a single step...

**Comment by SamVerschueren** [comment]
Yes switching entirely to Rollup by default wouldn't be a good thing. Especially because this means you would have to use `__moduleName` instead of `module.id` and compile to `system` instead of `commonjs`. Because all of this, it should be opt-in (if we are going for this).

I added the task below that performs the tree shaking for me in case you're interested.

```js
import * as path from 'path';
import * as gulp from 'gulp';
import * as gulpLoadPlugins from 'gulp-load-plugins';
import * as rollup from 'rollup';
import * as nodeResolve from 'rollup-plugin-node-resolve';
import * as commonjs from 'rollup-plugin-commonjs';
import * as uglify from 'rollup-plugin-uglify';
import * as toPromise from 'stream-to-promise';
import Config from '../../config';
import { templateLocals } from '../../utils';

const plugins = <any>gulpLoadPlugins();

const compileTemplate = () => {
	const stream = gulp.src(path.join(Config.TMP_DIR, '**/*.js'))
		.pipe(plugins.template(templateLocals()))
		.pipe(gulp.dest(Config.TMP_DIR));

	return toPromise(stream);
};

const treeshake = () => {
	return rollup
		.rollup({
			entry: path.join(Config.TMP_DIR, Config.BOOTSTRAP_FACTORY_PROD_MODULE),
			plugins: [
				nodeResolve({
					jsnext: true,
					module: true
				}),
				commonjs({
					include: 'node_modules/**'
				}),
				uglify()
			]
		})
		.then((bundle: any) => {
			return bundle.write({
				format: 'iife',
				dest: path.join(Config.JS_DEST, Config.JS_PROD_APP_BUNDLE)
			});
		});
};

export = (done: any) => {
	compileTemplate()
		.then(() => treeshake())
		.then(() => done())
		.catch((err: any) => done(err));
};
```

**Comment by mgechev** [comment]
Looks good! Let's think how it'll be best to add this as part of the seed and introduce bundling with Rollup in the next a couple of weeks :-).

**Comment by SamVerschueren** [comment]
Let me know if I can help you out with creating a PR if you know how you want this integrated!

**Comment by Perezmarc** [comment]
How is the implementation of tree shaking and lazy-loading in production mode going? 😃

**Comment by ogousa** [comment]
@mgechev I have problems with "lazy" branch (Chrome, Windows 10) while creating production build. 
First, I have to use 

**import { HomeComponent } from '../home/home.component';
import { HomeRoutingModule } from '../home/home-routing.module';**

instead of

**import { HomeComponent } from './home.component';
import { HomeRoutingModule } from './home-routing.module';**

The same trick for about module. Otherwise build.lazy-bundles.app.ts cannot find home folder:
![image](https://cloud.githubusercontent.com/assets/11968221/22832188/0fd0246c-ef7b-11e6-8625-cb4f40240640.png)

With these changes I can create bundles, but when I copy result to my web server I have this error:

![image](https://cloud.githubusercontent.com/assets/11968221/22832432/01395936-ef7c-11e6-8093-635db7ebf3ee.png)


Can you please help me?
Thanks.

**Comment by klesky** [comment]
hi am having the same error as @ogousa is having, anyone can shed more light into this?

**Comment by negberts** [comment]
Same here. Trying to get the Lazy Loading thing working but encountering the same errors (also when using the lazy branch).

**Comment by negberts** [comment]
@mgechev what settings for MINI_APP_MODULE and MINI_APP_BUNDLE are you using in this gist: https://gist.github.com/mgechev/63d374d51ac846797879a0cdc9c3b97c ?

**Comment by negberts** [comment]
@ogousa and @klesky did you get any further? Your problem is that in app.js the module is generated as "dist\\tmp\\app\\main.js" but is imported as "dist\tmp\app\main.js". Thats a problem voor SystemJs.
After I fixed that I received another error complaining about app.module and when I have a look in chrome to the source of app.module it shows me the content of index.html:

`<!DOCTYPE html><html lang="en"><head><base href="/"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><title>PIM Portal</title><meta name="description" content=""><meta name="viewport" content="width=device-width,initial-scale=1"><link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet"><link rel="stylesheet" href="/css/main.css?1496000579136"></head><body><pim-app>Loading...</pim-app><script>function module(){}</script><script src="/js/shims.js?1496000579131"></script><script src="/js/app.js?1496000579133"></script></body></html>`

Really strange...

**Comment by klesky** [comment]
@negberts i still could not find a solution for this

**Comment by negberts** [comment]
`"dist\\tmp\\app\\main.js"` but is imported as `"dist\tmp\app\main.js"`.

The slashes were removed in my comment. System.js sees \tmp as an escaped t, but the \ should be escaped by \\

**Comment by darecstowell** [comment]
Having the same issues as @ogousa and @klesky on the lazy branch. That branch could use a bit of love. Are there any plans to update it to Angular 4? My org is creating a fairly large app that benefits more from lazy loading than it does Tree Shaking/Rollup on master.
